### PR TITLE
Some fixes for S4MK2 controller and features

### PR DESCRIPTION
**Pull request by fayaaz**
Adds auto-slipmode and range for pitch fader as well

### CODE REVIEW

**Comment by Holzhaus** [comment]
Thanks. I can't review the code here (maybe @ywwg or @Be-ing?).

Can you rebase on 2.3? I don't think controller mapping changes need to go into the main branch. 
Also, If this adds new features or changes existing behaviour from the user's perpective, we need to document that on the wiki page. Please fork the https://github.com/mixxxdj/manual repository and document the changes in the corresponding file: https://github.com/mixxxdj/manual/blob/2.3/source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk3.rst

**Comment by JoergAtGithub** [comment]
Is AutoSlip-mode the same as the Flux-Mode in Traktor?
Did you try to use beatlooproll_activate to implement this behavior?

**Comment by Be-ing** [comment]
Yes please update the documentation. It is hard to review code when I'm not sure what it's intended to do.

**Comment by fayaaz** [comment]
@JoergAtGithub not sure what Flux-Mode is. I got the idea from Pioneer CDJs to activate the slip mode whenever something was going on (scratches, loops, beatrolls - although they already do that). Honestly I wrote it such a long time ago I've forgotten what it was meant for, but I use it at times and it kinda works.

@Be-ing ok - I should also mark this as WIP as there's some library selection stuff that seems to be broken with this controller on 2.3 that I'd like to fix.

**Comment by Holzhaus** [comment]
Corresponding manual PR: mixxxdj/manual #309

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js [hunk 1]
```diff
@@ -29,1266 +29,1322 @@ ShiftCueButtonAction = "REWIND";
...
+        TraktorS4MK2.linkChannelOutput("[Sampler7]", "play_indicator", TraktorS4MK2.outputChannelCallback);
+        TraktorS4MK2.linkChannelOutput("[Sampler8]", "play_indicator", TraktorS4MK2.outputChannelCallback);
+    } else if (TraktorS4MK2.RemixSlotButtonAction === "LOOPROLLS") {
+        TraktorS4MK2.linkDeckOutputs("beatlooproll_0.125_activate", TraktorS4MK2.outputCallback);
+        TraktorS4MK2.linkDeckOutputs("beatlooproll_0.25_activate", TraktorS4MK2.outputCallback);
+        TraktorS4MK2.linkDeckOutputs("beatlooproll_0.5_activate", TraktorS4MK2.outputCallback);
+        TraktorS4MK2.linkDeckOutputs("beatlooproll_1_activate", TraktorS4MK2.outputCallback);
+    }
+
+    engine.connectControl("[Recording]", "status", "TraktorS4MK2.onRecordingChanged");
```
Comment (Be-ing) [inline res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:567]:
```suggestion
    engine.makeConnect("[Recording]", "status", TraktorS4MK2.onRecordingChanged);
```
`engine.connectControl` is deprecated

**Comment by uklotzde** [comment]
What's the status of this PR? There are still code style issues. What about the corresponding manual PR https://github.com/mixxxdj/manual/pull/309?

We should try to get the controller mappings finalized and merged for 2.3.

**Comment by Holzhaus** [comment]
@fayaaz Can you fix the code style issues? Check the pre-commit check details for the diff.

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js [hunk 2]
```diff
@@ -20,7 +20,7 @@
 //    released playback will resume where the track would have been without the loop.
 // To choose sample launching, set the word in quotes below to SAMPLES, all caps.  For
 // rolls, change the word to LOOPROLLS (no space).
-RemixSlotButtonAction = "SAMPLES";
+RemixSlotButtonAction = "LOOPROLLS";
```
Comment (Be-ing) [inline res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:23]: Was this committed on accident?

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js [hunk 3]
```diff
@@ -1279,7 +1283,21 @@ TraktorS4MK2.callbackPregain = function(field) {
...
         }
-    } else {
+    }
+    else if (TraktorS4MK2.controller.play_shift_pressed === 1) {
+        var bpm_sensitivity_multipler = 50;
+        if (delta > 0) {
+            for (var i = 0; i < bpm_sensitivity_multipler; i++){
+                engine.setValue(group, "beats_adjust_faster", true);
+            }
+
```
Comment (Be-ing) [inline res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:1299]: extra newline

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js [hunk 4]
```diff
@@ -1709,6 +1727,38 @@ TraktorS4MK2.resetHandler = function(field) {
     }
 };
 
+TraktorS4MK2.pregainResetHandler = function(field) {
+
+
```
Comment (Be-ing) [inline res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:1737]: extra newlines

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js [hunk 5]
```diff
@@ -1709,6 +1727,38 @@ TraktorS4MK2.resetHandler = function(field) {
...
+    // Don't do something on push up
```
Comment (Be-ing) [inline res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:1738]: `// Don't so something on push release`

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js [hunk 6]
```diff
@@ -1709,6 +1727,38 @@ TraktorS4MK2.resetHandler = function(field) {
...
+    if (field.value === 0) {
+        return;
+    }
+    if (TraktorS4MK2.controller.play_shift_pressed) {
+        // If the 'play modifier' button is held down round the BPM
```
Comment (Be-ing) [inline res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js]:
```suggestion
         // If the 'play modifier' button is held down, round the BPM  
```

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js [hunk 7]
```diff
@@ -1709,6 +1727,38 @@ TraktorS4MK2.resetHandler = function(field) {
...
+    // Don't do something on push up
+    if (field.value === 0) {
+        return;
+    }
+    if (TraktorS4MK2.controller.play_shift_pressed) {
+        // If the 'play modifier' button is held down round the BPM   
+        var bpm = engine.getValue(group, 'bpm');
+        var rounded_value = Math.round(bpm);
+        var delta = (bpm - rounded_value)*100;
+        print(delta);
```
Comment (Be-ing) [inline res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js]: remove this line

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js [hunk 8]
```diff
@@ -1709,6 +1727,38 @@ TraktorS4MK2.resetHandler = function(field) {
...
+                engine.setValue(group, 'beats_adjust_slower', true)
+            }
+        }
+        else if (delta < 0){
+            for (var i = 0; i < delta; i++){
+                engine.setValue(group, 'beats_adjust_faster', true)
+            }
+        } 
+
+
```
Comment (Be-ing) [inline res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:1757]: extra new lines

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js (hunk 1 above)
Comment (Be-ing) [reply res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:571]: Why did you revert this?

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js (hunk 3 above)
Comment (Be-ing) [reply res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:1299]: oh sorry for the noise you fixed that in a later commit

Comment (Be-ing) [reply res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:1299]: err wait this extra line is still in the diff of the whole branch but the others are fixed

**Comment by Be-ing** [comment]
I don't care about eslint complaining about legacy code. The diff on this branch is already too big, let's not make it bigger.

**Comment by Be-ing** [comment]
ping @fayaaz there are only a few minor issues left to fix

File: res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js (hunk 1 above)
Comment (fayaaz) [reply res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:571]: Do you want me to update connectControl for all of them?

Comment (Be-ing) [reply res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js:571]: No, just the new code. If you want to clean up the rest in another PR after this, go for it. But this PR is already big enough.

**Comment by Be-ing** [comment]
Thanks for maintaining this mapping @fayaaz.

**Pull request merged**